### DISCUSSION TITLE
.union with string and object?

### DISCUSSION THREAD

**Discussion by danawoodman** [colinhacks/zod#2737] [Q&A] [answered]
I feel like I'm missing something, but shouldn't this code handle either a string _or_ an object with this shape?

```ts
const contact = z.object({
 		// other stuff here...
		created: z
			.union([
				z.string(),
				z.object({
					datetime: z.string(),
					by: z.string().optional(),
				}),
			])
			.optional(),
})
```

Instead of that, I get the error "expecting string received object"

I don't see any mention about this not being possible, I would expect Zod to accept either a string or this shaped object and only fail if a different value is passed.

What am I missing

**Comment by JacobWeisenburger** [comment, marked as answer]
Everthing seems to be working as expected:

```ts
const contact = z.object( {
    created: z.union( [
        z.string(),
        z.object( {
            datetime: z.string(),
            by: z.string().optional(),
        } ),
    ] ).optional(),
} )

console.log( contact.parse( { created: 'foo' } ) )
// { created: "foo" }

console.log( contact.parse( { created: { datetime: 'bar', by: 'baz' } } ) )
// { created: { datetime: "bar", by: "baz" } }
```

Please let me know if I'm missing something.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
